src/Components/Blazor/WebAssembly.Authentication/src/AuthenticationManager.cs

```diff
@@ -0,0 +1,446 @@
+            }
+        }
+
+        private async Task ProcessLogout(string returnUrl)
+        {
+            AuthenticationState.ReturnUrl = returnUrl;
+            var user = await AuthenticationService.GetCurrentUser();
+            var isauthenticated = user.Identity.IsAuthenticated;
+
+            if (isauthenticated)
+            {
+                var result = await AuthenticationService.SignOutAsync(new RemoteAuthenticationContext<TAuthenticationState> { State = AuthenticationState });
+                switch (result.Status)
+                {
+                    case RemoteAuthenticationStatus.Redirect:
+                        break;
+                    case RemoteAuthenticationStatus.Success:
+                        await NavigateToReturnUrl(returnUrl);
+                        break;
+                    case RemoteAuthenticationStatus.OperationCompleted:
+                        break;
+                    case RemoteAuthenticationStatus.Failure:
+                        var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogoutFailedPath}?message={result.ErrorMessage}").ToString();
+                        Navigation.NavigateTo(uri);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Invalid authentication result status '{result.Status ?? "(null)"}'.");
+                }
+            }
+        }
+
+        private async Task ProcessLogoutCallback()
+        {
+            var result = await AuthenticationService.CompleteSignOutAsync(new RemoteAuthenticationContext<TAuthenticationState> { Url = Navigation.Uri });
+            switch (result.Status)
+            {
+                case RemoteAuthenticationStatus.Redirect:
+                    // There should not be any redirects as the only time completeAuthentication finishes
+                    // is when we are doing a redirect sign in flow.
+                    throw new InvalidOperationException("Should not redirect.");
+                case RemoteAuthenticationStatus.Success:
+                    await NavigateToReturnUrl(GetReturnUrl(result.State, Navigation.ToAbsoluteUri(ApplicationPaths.LogoutSucceededPath).ToString()));
+                    break;
+                case RemoteAuthenticationStatus.OperationCompleted:
+                    break;
+                case RemoteAuthenticationStatus.Failure:
+                    var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogoutFailedPath}?message={result.ErrorMessage}").ToString();
+                    Navigation.NavigateTo(uri);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Invalid authentication result status '{result.Status ?? "(null)"}'.");
+            }
+        }
+
+        private string GetReturnUrl(TAuthenticationState state, string defaultReturnUrl = null)
+        {
+            var fromQuery = GetParameter("returnUrl");
```

Comment:
If `state.ReturnUrl` is non-null, could we short-circuit and return that without going through the querystring parsing and validating work?

Reply:
done